### PR TITLE
Allow op processing in paused state

### LINKED ISSUE #5477: Apply ops in ODSP snapshot before returning container from Container.load()
Need to figure out how to apply ops that are coming  in snapshot, and do so in generic way not to force other drivers to follow ODSP model.

Some key bugs describing why it's pretty broken for users:
https://office.visualstudio.com/OC/_workitems/edit/4899495
https://office.visualstudio.com/OC/_workitems/edit/4899481

### PR DESCRIPTION
**Pull request by vladsud**
This is first (and possible the only) step to resolve https://github.com/microsoft/FluidFramework/issues/5477
Please see breaking.md - there is substantial change in behavior.

The reason I believe it's good direction - it allows ops to trickle in sooner, including from storage (we make proactive request to storage as part of booting DeltaManager). I think it was original behavior, but our oldest release (0.10) already has pause/resume implemented.

It may expose bugs in existing code that may incorrectly assume that state of the universe does not change through initial loading / rendering. These bugs would show up in any other place where we would do virtualization (delay loading components). And to some extent, that's already what happens even today in Bohemia if we hit timeout and resume() is called before rendering is done. That said, if any bugs exists in existing components, with this change chances of hitting them goes way up.

Alternative is to apply ops before Container.load() returns, but pause right before returning. I do not like that direction, but it's an option.

Feedback?

### CODE REVIEW

**Review by christiango — APPROVED** [review]
Are there any tests we can add to validate this behavior more thoroughly?

**Comment by curtisman** [comment]
What is pause needed for?

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -771,6 +770,9 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
             // Propagate current connection state through the system.
             this.propagateConnectionState();
             if (!this.closed) {
+                this._deltaManager.inbound.resume();
```
Comment (christiango) [inline packages/loader/container-loader/src/container.ts:773]: So I guess we don't "block" anything on processing the ops. Is that correct? So when I get the container back from the loader, the ops may not have been applied yet, right?

Comment (vladsud) [reply packages/loader/container-loader/src/container.ts:773]: That is correct. I'm looking at it as multi-stage work. First - ensure that ops can trickle through as quickly as possible. This ensures that if ops are coming from storage (not cached in any form), they make it as soon as possible to end result.
The there potential fix - figure out how to guarantee that ops that are cached applied before  container returns.

**Comment by vladsud** [comment]
> What is pause needed for?
That's poor men's implementation of prioritization.
Loading bundle with socket.io takes a bunch of CPU resourced from critical work. 
Fetching extra tokens used to be slow as well (not sure if it's still the case).
And initial socket.io handshake (couple messages on socket) were also very expensive (likely code parsing / JITting?)

So pausing is really about moving this cost off critical path, nothing more. At least as it is used today in Bohemia.
Clearly there might be other cases where you might want to get non-changing container.
I think that should be achieved through using fixed version - I believe I should account for that in this PR.
Otherwise, if you pass nothing (undefined), that means (has semantic) of latest state of the file.

**Comment by vladsud** [comment]
> Are there any tests we can add to validate this behavior more thoroughly?

You mean on Bohemia side? I believe most of the test code in FF repo does not use paused container, so it was always a default flow.
Hard to answer that question. I think the easiest to force it - load existing files (or test files) from old snapshot and let ops flow in as we are finishing boot / render. It's clearly hit-and-miss in terms of guaranteeing to find a bug (as such bugs would be intermittent), but if done many times in a row, it's a good test to validate correctness of the code.

Note that this is not specific to boot. Same is true for delay-loading components. So other way to test it - having a test case with 2 clients, where one client submits a ton of ops non-stop for given component (or multiple components), and another client loads such component(s) while ops are flowing.

**Comment by christiango** [comment]
> > Are there any tests we can add to validate this behavior more thoroughly?
> 
> You mean on Bohemia side? I believe most of the test code in FF repo does not use paused container, so it was always a default flow.
> Hard to answer that question. I think the easiest to force it - load existing files (or test files) from old snapshot and let ops flow in as we are finishing boot / render. It's clearly hit-and-miss in terms of guaranteeing to find a bug (as such bugs would be intermittent), but if done many times in a row, it's a good test to validate correctness of the code.
> 
> Note that this is not specific to boot. Same is true for delay-loading components. So other way to test it - having a test case with 2 clients, where one client submits a ton of ops non-stop for given component (or multiple components), and another client loads such component(s) while ops are flowing.

I would have expected tests on the FF side to encode the intended behavior of the pause flag. Like ensuring that local ops are not blocked.

**Comment by curtisman** [comment]
Should remove or rename "pause" all together?  May it is just a mode to connect `disconnected` until user explicit call `connect` instead of `pause`, which now have unclear semantics.

**Comment by anthony-murphy** [comment]
should be easy to write end to end test that validate at least some of the pause behavior, as we spawn containers, and load from summaries already, not reason you can't also pass the pause header

**Comment by vladsud** [comment]
99% of our tests go through non-pause/resume workflow - with this change I'm changing Bohemia experience to be closer to what we actually test in FF repo (not exactly, but the problem that I fear the most are the same - ops trickling in while initial boot happens and components not being prepared for it). There is not much more testing I see here happening in FF, as area where we are likely to see problems are in "user" code - components that expect the world to not change through load sequence.

We do have a test case to validate pause / resume() workflow - "loaded container is paused using loader pause flags".

The case that is missing - validating that we "catch up" to some reasonable latest state in paused state (i.e. actual change in behavior - processing all trailing ops that were known at the time of boot). That one needs to be added.

> Should we remove or rename "pause" all together? 

In general, I think it's useful for case of loading from specific version, as it clearly says - I want that state, and not some other state. If you do not specify version, then I'd say semantically, it's Ok (not perfect, but Ok). In a sense that we are not connecting to ordering service and container is stuck at some state. It's Ok (not great) because it may take a while for container to get to that state (fetching ops and processing it). But it fulfils overall desire of getting "current" state of file and stop there.

So I believe two additions are needed:
1. Not process any ops if loaded from a version
2. Test that we process all known ops at the time of boot (from latest).

Please let me know if I'm missing something or you disagree with my assessment

**Comment by vladsud** [comment]
Superseded by https://github.com/microsoft/FluidFramework/pull/5773

**Pull request closed without merging**